jbmc/src/java_bytecode/select_pointer_type.cpp

```diff
@@ -25,61 +25,60 @@ pointer_typet select_pointer_typet::convert_pointer_type(
   // a generic parameter, specialize it with concrete types
   if(!generic_parameter_specialization_map.empty())
   {
-    generic_parameter_recursion_trackingt visited;
-    const auto &type = specialize_generics(
-      pointer_type, generic_parameter_specialization_map, visited);
-    INVARIANT(visited.empty(), "recursion stack must be empty here");
-    return type;
+    return specialize_generics(
+      pointer_type, generic_parameter_specialization_map);
   }
   else
   {
     return pointer_type;
   }
 }
 
+/// Return the first concrete type instantiation if any such exists.
+///
+/// Example:
+/// `class MyGeneric<T,U> { MyGeneric<U,T> gen; T t;}`
+/// When instantiating `MyGeneric<Integer,String> my` we need to for example
+/// resolve the type of `my.gen.t`. The map would in this context contain
+/// - T -> (Integer, U)
+/// - U -> (String, T)
+///
+/// Note that the top of the stacks for T and U create a recursion T -> U,
+/// U-> T. We want to break it and specialize `T my.gen.t` to `String my.gen.t`.
+/// \param pointer_type: The pointer type to specialize
+/// \param generic_parameter_specialization_map: Map of type names to
+///   specialization stack
+/// \return The first instantiated type for the generic type or nothing if no
+///   such instantiation exists.
```

Comment:
This should be updated slightly - it sounds l like it's only handling the recursion case (because it's just moved from the function that was breaking recursion). Also, move the documentation to the header.